src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -20,45 +20,171 @@
+            boolean isResultSet = setResultFromExpression();
+            if (isResultSet) {
+                return;
+            }
+        }
+
+        while (_child.hasNext()) {
+            _inputTuple = _child.next();
+            _tupleContext.removeAllVariables();             // clear the previous variables
+            _tupleContext.setBindingsFromTuple(_inputTuple);      // assign new variables from new tuple
+
+            _expression.open(_tupleContext);
+            boolean isResultSet = setResultFromExpression();
+            if (isResultSet) {
+                return;
+            }
+        }
+
+        // execution reaches here when there are no more results
+        _child.close();
+        this._hasNext = false;
+    }
+
+    /**
+     * _expression has to be open prior to call.
+     * @return true if _nextLocalTupleResult is set and _hasNext is true, false otherwise
+     */
+    private boolean setResultFromExpression() {
+        if (_expression.hasNext()) {     // if expression returns a value, set it as next
+            List<Item> results = new ArrayList<>();
+            results.add(_expression.next());
+            FlworTuple newTuple = new FlworTuple(_inputTuple, _variableName, results);
+            _nextLocalTupleResult = newTuple;
+            this._hasNext = true;
+            return true;
+        } else {
+            _expression.close();
+            return false;
+        }
+    }
+
+    private void initialFor_setNextLocalTupleResult() {
+        if (_expression.hasNext()) {
+            List<Item> results = new ArrayList<>();
+            results.add(_expression.next());
+            FlworTuple newTuple = new FlworTuple(_variableName, results);
+            _nextLocalTupleResult = newTuple;
+            this._hasNext = true;
+        } else {
+            _expression.close();
+            this._hasNext = false;
+        }
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) { //if it's not a start clause
+            _child.open(_currentDynamicContext);
+            _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+
+            setNextLocalTupleResult();
+
+        } else {    //if it's a start clause
+            _expression.open(this._currentDynamicContext);
+
+            initialFor_setNextLocalTupleResult();
+        }
+    }
+
+    @Override
+    public void close() {
+        if (_child != null) {
+            _child.close();
+        }
     }
 
 
     @Override
     public JavaRDD<FlworTuple> getRDD() {
-        if (this._rdd == null) {
-            String variableReference = ((VariableReferenceIterator)this._children.get(0)).getVariableName();
-            RuntimeIterator assignmentExpression = this._children.get(1);
+        if (this._rdd == null) {        //if it's a start clause
```

Comment:
This if should be removed, because the rdd shouldn't be cached. It may change across different outer dynamic context (at some point we will pass it as a parameter to getRDD() as we do for items).

Reply:
Should we remove the RDD caching in the getRDD function of letClause as well?

Reply:
yes, it applies there too.